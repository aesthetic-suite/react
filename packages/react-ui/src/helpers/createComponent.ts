/* eslint-disable @typescript-eslint/no-explicit-any */

import { forwardRef } from 'react';
import { CommonHtmlProps, HtmlElementType, OmitUnwantedHtmlProps } from '../types';

export function createComponent<Props extends object, Element extends HTMLElement = HTMLElement>(
	component: React.ForwardRefRenderFunction<Element, Props>,
) {
	return forwardRef(component);
}

export type DynamicProps<
	As extends React.ElementType,
	Props extends object
> = OmitUnwantedHtmlProps<React.ComponentPropsWithRef<As>> &
	Props & {
		/**
		 * Native HTML element or custom component to render as.
		 */
		as?: As;
	};

// This is based on the `React.ExoticComponent` but supports the dynamic `as` prop
// through a generic on the render function. We want the props on the consumer to change
// based on the `as` prop, but we *do not* want to type the props/ref in the implementation,
// as it makes the types too complex and we spread them into the DOM regardless.
export interface DynamicComponent<Props extends object, As extends React.ElementType> {
	<T extends As>(props: DynamicProps<T, Props>): React.ReactElement | null;
	readonly $$typeof: symbol;
	displayName?: string;
}

// We only want to return generic attributes when an HTML element,
// otherwise return the props from the custom component.
export type InferAsProps<T> = T extends HtmlElementType
	? CommonHtmlProps
	: T extends React.JSXElementConstructor<infer P>
	? P
	: {};

export function createDynamicComponent<
	Props extends object,
	As extends React.ElementType = HtmlElementType
>(component: React.ForwardRefRenderFunction<any, InferAsProps<As> & Props & { as?: As }>) {
	return (createComponent(component) as unknown) as DynamicComponent<Props, As>;
}
