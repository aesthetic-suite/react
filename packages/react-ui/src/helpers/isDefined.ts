export function isDefined(...values: unknown[]): boolean {
	return values.some((value) => typeof value !== 'undefined' && value !== null && value !== false);
}
