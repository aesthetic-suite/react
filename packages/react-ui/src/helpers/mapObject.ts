import { objectLoop } from '@aesthetic/utils';

export function mapObject<P, N>(
  base: Record<string, P>,
  onValue: (value: P) => N | undefined,
  onKey?: (key: string) => string,
): Record<string, N> {
  const object: Record<string, unknown> = {};

  objectLoop(base, (value, key) => {
    const nextValue = onValue(value);

    if (nextValue !== undefined) {
      object[onKey ? onKey(key) : key] = nextValue;
    }
  });

  return object as Record<string, N>;
}
