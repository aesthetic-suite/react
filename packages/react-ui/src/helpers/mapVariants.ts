import { ElementStyles } from '@aesthetic/react';
import { arrayLoop } from '@aesthetic/utils';

export function mapVariants<T>(
	type: string,
	enums: T[],
	factory: (value: T, typeName: string) => ElementStyles,
) {
	const object: Partial<Record<string, ElementStyles>> = {};

	arrayLoop(enums, (value) => {
		object[`${type}:${value}`] = factory(value, type);
	});

	return object as Record<string, ElementStyles>;
}
