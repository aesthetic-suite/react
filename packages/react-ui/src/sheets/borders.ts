import { style } from '@aesthetic/react';

export const borderStyleSheet = style(
	(css) => ({
		borderColor: 'transparent',
		borderStyle: 'solid',

		'@variants': {
			'border:sm': {
				borderWidth: css.var('border-sm-width'),
				borderRadius: css.var('border-sm-radius'),
			},
			'border:df': {
				borderWidth: css.var('border-df-width'),
				borderRadius: css.var('border-df-radius'),
			},
			'border:lg': {
				borderWidth: css.var('border-lg-width'),
				borderRadius: css.var('border-lg-radius'),
			},
		},
	}),
	'borders',
);
