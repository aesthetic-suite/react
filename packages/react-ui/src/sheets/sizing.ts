import { style } from '@aesthetic/react';

export const sizingStyleSheet = style(
	(css) => ({
		'@variants': {
			'size:sm': {
				lineHeight: css.var('text-sm-line-height'),
				fontSize: css.var('text-sm-size'),
				padding: css.unit(0.25, 0.75),
				minWidth: 4 * css.tokens.spacing.unit,
			},
			'size:df': {
				lineHeight: css.var('text-df-line-height'),
				fontSize: css.var('text-df-size'),
				padding: css.unit(0.75, 1.25),
				minWidth: 5 * css.tokens.spacing.unit,
			},
			'size:lg': {
				lineHeight: css.var('text-lg-line-height'),
				fontSize: css.var('text-lg-size'),
				padding: css.unit(1.25, 2),
				minWidth: 6 * css.tokens.spacing.unit,
			},
		},
	}),
	'sizing',
);
