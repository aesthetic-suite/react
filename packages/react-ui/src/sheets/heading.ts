import { HEADING_LEVELS, style } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';

export const headingStyleSheet = style(
	(css) => ({
		fontFamily: css.var('typography-font-heading'),

		'@variants': {
			...mapVariants('level', HEADING_LEVELS, (level) => ({
				lineHeight: css.var(`heading-l${level}-line-height`),
				fontSize: css.var(`heading-l${level}-size`),
				letterSpacing: css.var(`heading-l${level}-letter-spacing`),
			})),
		},
	}),
	'typography',
);
