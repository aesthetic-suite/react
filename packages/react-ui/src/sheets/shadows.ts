import { SHADOW_SIZES, style } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';

export const shadowStyleSheet = style(
	(css) => ({
		'@variants': {
			...mapVariants('shadow', SHADOW_SIZES, (size) => ({
				boxShadow: [
					css.var(`shadow-${size}-x`),
					css.var(`shadow-${size}-y`),
					css.var(`shadow-${size}-blur`),
					css.var(`shadow-${size}-spread`),
					'rgba(0, 0, 0, .8)',
				].join(' '),
			})),
		},
	}),
	'shadows',
);
