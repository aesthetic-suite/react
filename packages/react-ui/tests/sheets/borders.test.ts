import { renderStyleSheet } from '@aesthetic/react';
import { borderStyleSheet } from '../../src/sheets/borders';
import { getRenderedStyles, withStyles } from '../helpers';

describe('borders', () => {
	it(
		'renders styles',
		withStyles(() => {
			renderStyleSheet(borderStyleSheet);

			expect(getRenderedStyles('standard')).toMatchSnapshot();
		}),
	);
});
