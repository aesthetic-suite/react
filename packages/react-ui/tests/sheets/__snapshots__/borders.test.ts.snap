// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`borders renders styles 1`] = `".cfdmyha {border-style: solid;}.c1e95qeq {border-width: var(--border-sm-width);}.c18pn6tu {border-radius: var(--border-sm-radius);}.c1q9bcr2 {border-width: var(--border-df-width);}.c1rooa26 {border-radius: var(--border-df-radius);}.c1xgoolj {border-width: var(--border-lg-width);}.cmt9y9z {border-radius: var(--border-lg-radius);}"`;
